pyomo/contrib/gdpopt/nlp_solve.py

```diff
@@ -311,19 +312,17 @@ def detect_unfixed_discrete_vars(model):
 
 def preprocess_subproblem(m, config):
     """Applies preprocessing transformations to the model."""
-    # fbbt(m, integer_tol=config.integer_tolerance)
+    # First do FBBT
+    fbbt(m, integer_tol=config.integer_tolerance,
+         feasibility_tol=config.constraint_tolerance)
     xfrm = TransformationFactory
-    xfrm('contrib.propagate_eq_var_bounds').apply_to(m)
+    # Now that we've tightened bounds, see if any variables are fixed because
+    # their lb is equal to the ub (within tolerance)
     xfrm('contrib.detect_fixed_vars').apply_to(
-        m, tolerance=config.variable_tolerance)
-    xfrm('contrib.propagate_fixed_vars').apply_to(m)
+         m, tolerance=config.variable_tolerance)
+    # Now, if something got fixed to 0, we might have 0*var terms to remove
     xfrm('contrib.remove_zero_terms').apply_to(m)
-    xfrm('contrib.propagate_zero_sum').apply_to(m)
-    xfrm('contrib.constraints_to_var_bounds').apply_to(
-        m, tolerance=config.variable_tolerance)
-    xfrm('contrib.detect_fixed_vars').apply_to(
-        m, tolerance=config.variable_tolerance)
-    xfrm('contrib.propagate_zero_sum').apply_to(m)
+    # Last, check if any constraints are now trivial and deactivate them
```

Comment:
See main review's comment.

Reply:
@bernalde FBBT does handle the `propagate_zero_sum` transformation:

```
In [1]: import pyomo.environ as pe

In [2]: m = pe.ConcreteModel()

In [3]: m.a = pe.Set(initialize=list(range(5)))

In [4]: m.x = pe.Var(m.a, bounds=(0, None))

In [5]: m.c1 = pe.Constraint(expr=sum(m.x.values()) == 0)

In [6]: from pyomo.contrib.fbbt.fbbt import fbbt

In [7]: fbbt(m)
Out[7]: <pyomo.common.collections.component_map.ComponentMap at 0x7fc4e10e40a0>

In [8]: m.x.pprint()
x : Size=5, Index=a
    Key : Lower : Value : Upper : Fixed : Stale : Domain
      0 :     0 :  None :   0.0 : False :  True :  Reals
      1 :     0 :  None :   0.0 : False :  True :  Reals
      2 :     0 :  None :   0.0 : False :  True :  Reals
      3 :     0 :  None :   0.0 : False :  True :  Reals
      4 :     0 :  None :   0.0 : False :  True :  Reals
```

These variables will then get fixed in the `detect_fixed_vars` transformation. FBBT will work on more general constraints as well:

```
In [9]: m.x.setub(None)

In [10]: m.x.pprint()
x : Size=5, Index=a
    Key : Lower : Value : Upper : Fixed : Stale : Domain
      0 :     0 :  None :  None : False :  True :  Reals
      1 :     0 :  None :  None : False :  True :  Reals
      2 :     0 :  None :  None : False :  True :  Reals
      3 :     0 :  None :  None : False :  True :  Reals
      4 :     0 :  None :  None : False :  True :  Reals

In [11]: del m.c1

In [12]: m.c1 = pe.Constraint(expr=sum(i*m.x[i] for i in m.a) == 0)

In [13]: fbbt(m)
Out[13]: <pyomo.common.collections.component_map.ComponentMap at 0x7fc4e1cad730>

In [14]: m.x.pprint()
x : Size=5, Index=a
    Key : Lower : Value : Upper : Fixed : Stale : Domain
      0 :     0 :  None :  None : False :  True :  Reals
      1 :     0 :  None :   0.0 : False :  True :  Reals
      2 :     0 :  None :   0.0 : False :  True :  Reals
      3 :     0 :  None :   0.0 : False :  True :  Reals
      4 :     0 :  None :   0.0 : False :  True :  Reals
```